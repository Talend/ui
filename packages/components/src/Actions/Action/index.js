import Action, { wrapOnClick } from './Action.component';

export { wrapOnClick };
export default Action;
