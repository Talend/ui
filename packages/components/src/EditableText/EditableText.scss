@import '~@talend/bootstrap-theme/src/theme/animation';

@mixin ellipsis {
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
}

$tc-icon-editable-text-size: 0.8rem !default;
$tc-circle-editable-text-size: 16px !default;

:global(.tc-editable-text-blink) {
	@include heartbeat(object-blink);
}

.tc-editable-text {
	margin: 0 110px 0 0;
	max-width: 90rem;

	&-title {
		display: flex;
		align-items: center;
		max-width: 90rem;

		:global(.tc-action-disabled-btn-container) {
			height: 19px;
		}
	}

	&-wording {
		@include ellipsis;
		flex: 0 1 auto;
	}

	&-pencil {
		padding-left: $padding-small;
		padding-right: $padding-small;
		flex: 0 0 auto;
		line-height: inherit;
		min-height: inherit;

		&:global(.btn-link) {
			color: $silver;
		}
	}

	& &-form {
		position: relative;
		width: 100%;
		display: flex;
		align-items: center;
		padding-right: $padding-large;

		:global(.form-group) {
			width: 100%;
			margin: 0;
		}

		& .tc-editable-text-form-input {
			width: 100%;
			padding-right: 2 * $tc-circle-editable-text-size + $padding-large;
			&::selection {
				background: lighten($scooter, 25%);
			}
		}

		&-buttons {
			display: flex;
			position: absolute;
			right: $padding-large;
			align-items: center;
			justify-content: center;
			height: 32px;
			top: 0;

			&-icon {
				border: 0.1rem solid $dark-silver;
				color: $dark-silver;
				width: $tc-circle-editable-text-size;
				height: $tc-circle-editable-text-size;
				min-height: auto;
				display: flex;
				border-radius: 50%;
				padding: 0;
				justify-content: center;
				align-items: center;

				> :global(.tc-svg-icon) {
					width: $tc-icon-editable-text-size;
					height: $tc-icon-editable-text-size;
				}

				&:last-child {
					margin-left: $padding-small;
				}
			}

			&-submit {
				border-color: $rio-grande;
				background: $rio-grande;
				fill: white;

				> svg {
					color: white;
				}
			}
		}

		:global(.form-group) {
			padding-top: 0;
		}
	}
}
