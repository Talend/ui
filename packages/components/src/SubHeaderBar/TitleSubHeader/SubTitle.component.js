import PropTypes from 'prop-types';

import Skeleton from '../../Skeleton';
import titleSubHeaderCssModule from './TitleSubHeader.module.scss';
import { getTheme } from '../../theme';

const theme = getTheme(titleSubHeaderCssModule);

function DefaultSubTitle({ subTitle, subTitleProps }) {
	return (
		<small className={theme('tc-subheader-details-text-subtitle')} {...subTitleProps}>
			{subTitle}
		</small>
	);
}

DefaultSubTitle.propTypes = {
	subTitle: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
	subTitleProps: PropTypes.object,
};

export function SubTitle({
	subTitleLoading,
	subTitle,
	subTitleAs: SubTitleAs = DefaultSubTitle,
	...rest
}) {
	if (subTitleLoading) {
		return (
			<Skeleton
				className={theme('tc-subheader-details-loading-subtitle')}
				type={Skeleton.TYPES.text}
				size={Skeleton.SIZES.large}
			/>
		);
	}

	if (subTitle) {
		return <SubTitleAs subTitle={subTitle} {...rest} />;
	}

	return null;
}

SubTitle.propTypes = {
	subTitle: PropTypes.node,
	subTitleLoading: PropTypes.bool,
	subTitleAs: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};
