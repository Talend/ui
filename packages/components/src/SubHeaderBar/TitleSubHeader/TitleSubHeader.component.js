import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Skeleton from '../../Skeleton';
import { EditableText } from '../..';
import theme from './TitleSubHeader.scss';
import Icon from '../../Icon';
import Inject from '../../Inject';
import getDefaultT from '../../translate';
import TooltipTrigger from '../../TooltipTrigger';

function TitleSubHeader({
	title,
	iconId,
	loading,
	inProgress,
	editable,
	getComponent,
	onEdit,
	onCancel,
	onSubmit,
	...rest
}) {
	const [isEditMode, setIsEditMode] = React.useState(false);
	function handleEdit() {
		setIsEditMode(true);
		if (onEdit) {
			onEdit();
		}
	}
	function handleCancel() {
		setIsEditMode(false);
		if (onCancel) {
			onCancel();
		}
	}
	function handleSubmit() {
		setIsEditMode(false);
		if (onSubmit) {
			onSubmit();
		}
	}
	if (loading) {
		return <Skeleton type={Skeleton.TYPES.text} size={Skeleton.SIZES.large} />;
	}
	const InjectedEditableText = Inject.get(getComponent, 'EditableText', EditableText);
	return (
		<div
			className={classNames(theme['tc-subheader-details'], 'tc-subheader-details', {
				[theme['tc-subheader-details-blink']]: inProgress,
				'tc-subheader-details-blink': inProgress,
			})}
		>
			{iconId && (
				<Icon
					name={iconId}
					className={classNames(theme['tc-subheader-details-icon'], 'tc-subheader-details-icon')}
				/>
			)}
			<div className={classNames(theme['tc-subheader-details-text'], 'tc-subheader-details-text')}>
				<div
					className={classNames(
						theme['tc-subheader-details-text-title'],
						'tc-subheader-details-text-title',
					)}
				>
					{editable ? (
						<InjectedEditableText
							text={title}
							inProgress={inProgress}
							feature="subheaderbar.rename"
							componentClass="h1"
							onEdit={handleEdit}
							onCancel={handleCancel}
							onSubmit={handleSubmit}
							{...rest}
						/>
					) : (
						<TooltipTrigger label={title} tooltipPlacement="bottom">
							<h1
								className={classNames(
									theme['tc-subheader-details-text-title-wording'],
									'tc-subheader-details-text-title-wording',
								)}
							>
								{title}
							</h1>
						</TooltipTrigger>
					)}
				</div>
				{!isEditMode ? <SubTitle {...rest} /> : null}
			</div>
		</div>
	);
}

function SubTitle({ subTitleLoading, subTitle }) {
	if (subTitleLoading) {
		return (
			<Skeleton
				className={classNames(
					theme['tc-subheader-details-loading-subtitle'],
					'tc-subheader-details-loading-subtitle',
				)}
				type={Skeleton.TYPES.text}
				size={Skeleton.SIZES.large}
			/>
		);
	}

	if (subTitle) {
		return (
			<small
				className={classNames(
					theme['tc-subheader-details-text-subtitle'],
					'tc-subheader-details-text-subtitle',
				)}
			>
				{subTitle}
			</small>
		);
	}

	return null;
}

SubTitle.propTypes = {
	subTitle: PropTypes.string,
	subTitleLoading: PropTypes.bool,
};

TitleSubHeader.propTypes = {
	title: PropTypes.string.isRequired,
	iconId: PropTypes.string,
	loading: PropTypes.bool,
	inProgress: PropTypes.bool,
	editable: PropTypes.bool,
	subTitle: PropTypes.string,
	onEdit: PropTypes.func,
	onSubmit: PropTypes.func,
	onCancel: PropTypes.func,
	...Inject.PropTypes,
};

TitleSubHeader.defaultProps = {
	loading: false,
	inProgress: false,
	t: getDefaultT(),
	editable: false,
};

export { TitleSubHeader as default, SubTitle };
