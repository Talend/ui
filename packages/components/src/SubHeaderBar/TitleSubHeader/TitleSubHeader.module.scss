@use '~@talend/bootstrap-theme/src/theme/guidelines' as *;
@use '~@talend/design-tokens/lib/tokens';

@import '~@talend/bootstrap-theme/src/theme/animation';

@mixin ellipsis {
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
}

$tc-input-subheader-size-large: 2.4rem !default;
$tc-input-subheader-size-medium: 1.4rem !default;
$header-title-max-width: 90rem;

:global(.tc-subheader-details-blink) {
	@include heartbeat(object-blink);
}

.tc-subheader-details {
	display: flex;
	margin: 0;
	height: 100%;

	& &-icon {
		align-self: center;
		width: $tc-input-subheader-size-large;
		height: $tc-input-subheader-size-large;
		color: tokens.$coral-color-accent-icon;
		margin-right: $padding-large;
	}

	& &-text {
		display: flex;
		flex-direction: column;
		justify-content: center;
		overflow: hidden;
		flex: 1 auto;
		margin-right: 110px;
		max-width: $header-title-max-width;
		align-items: flex-start;

		&-title {
			display: inline-flex;
			align-items: center;
			width: 100%;

			// force InlineEditing max space to allow ellipsis
			> * {
				max-width: $header-title-max-width;
			}

			&-wording,
			&-wording-button {
				@include ellipsis;
				line-height: unset;
				margin: 0;
			}

			input {
				width: tokens.$coral-sizing-maximal;
			}
		}

		&-subtitle {
			@include ellipsis;
			color: tokens.$coral-color-neutral-text;
			font-size: $tc-input-subheader-size-medium;
		}
	}

	& &-loading-subtitle {
		margin-top: $padding-smaller;
	}
}
