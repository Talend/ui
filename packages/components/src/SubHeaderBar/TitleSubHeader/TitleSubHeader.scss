@import '~@talend/bootstrap-theme/src/theme/animation';

@mixin ellipsis {
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
}

$tc-input-subheader-size-large: 2.4rem !default;
$tc-input-subheader-size-medium: 1.4rem !default;

:global(.tc-subheader-details-blink) {
	@include heartbeat(object-blink);
}

.tc-subheader-details {
	display: flex;
	margin: 0;
	height: 100%;

	&-icon {
		align-self: center;
		width: $tc-input-subheader-size-large;
		height: $tc-input-subheader-size-large;
		color: $brand-primary;
		margin-right: $padding-large;
	}

	& &-text {
		display: flex;
		flex-direction: column;
		justify-content: center;
		overflow: hidden;
		flex: 1 auto;
		margin-right: 110px;
		max-width: 90rem;
		align-items: flex-start;

		&-title {
			display: inline-flex;
			align-items: center;

			&-wording,
			&-wording-button {
				@include ellipsis;
				line-height: unset;
				margin: 0;
			}

			:global {
				.tc-editable-text-wording {
					line-height: unset;
					margin: 0;
				}

				.tc-editable-text-form-input {
					width: 30rem;
				}

				.tc-editable-text-form-buttons-submit:hover {
					background: $scooter;

					& svg {
						background: $scooter;
					}
				}

				.tc-editable-text-pencil {
					opacity: 0;

					&:focus,
					&:active {
						opacity: 1;
					}
				}

				:hover {
					.tc-editable-text-pencil {
						opacity: 1;
					}
				}
			}
		}

		&-subtitle {
			@include ellipsis;
			color: $black;
			font-size: $tc-input-subheader-size-medium;
		}
	}

	& &-loading-subtitle {
		margin-top: $padding-smaller;
	}
}
