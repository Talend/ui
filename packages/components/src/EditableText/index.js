import EditableText from './EditableText.component';

export default EditableText;
