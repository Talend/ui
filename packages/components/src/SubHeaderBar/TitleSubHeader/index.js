import TitleSubHeader from './TitleSubHeader.component';

export default TitleSubHeader;
