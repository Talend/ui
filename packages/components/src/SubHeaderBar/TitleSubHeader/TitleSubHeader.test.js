import React from 'react';
import { shallow } from 'enzyme';
import { Label } from 'react-bootstrap';
import Skeleton from '../../Skeleton';
import Icon from '../../Icon';
import Action from '../../Actions/Action';
import EditableText from '../../EditableText';
import TitleSubHeader, { SubTitle } from './TitleSubHeader.component';

describe('TitleSubHeader', () => {
	let defaultProps;
	beforeEach(
		() =>
			(defaultProps = {
				title: 'myTitle',
				onEdit: jest.fn(),
				onSubmit: jest.fn(),
			}),
	);
	it('should render', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} iconId="myIconId" />);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render Icon', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} iconId="myIconId" />);
		expect(wrapper.find(Icon)).toHaveLength(1);
		expect(wrapper.find(Icon).get(0).props.name).toEqual('myIconId');
	});
	it('should not render Icon', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} />);
		expect(wrapper.find(Icon)).toHaveLength(0);
	});
	it('should render plain text title', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} editable={false} />);
		expect(wrapper.find(TitleSubHeader)).toHaveLength(0);
		expect(wrapper.find('h1')).toHaveLength(1);
	});
	it('should render EditableText', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} editable />);
		expect(wrapper.find(EditableText)).toHaveLength(1);
		expect(wrapper.find(EditableText).get(0).props.feature).toBe('subheaderbar.rename');
		expect(wrapper.find('h1')).toHaveLength(0);
	});
	it('should render skeleton', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} loading />);
		expect(wrapper.find(Skeleton)).toHaveLength(1);
	});
	it('should render inProgress', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} inProgress />);
		expect(wrapper.props().className).toEqual(
			'tc-subheader-details theme-tc-subheader-details tc-subheader-details-blink theme-tc-subheader-details-blink',
		);
	});
});

describe('TitleSubHeader', () => {
	let defaultProps;
	beforeEach(() => {
		defaultProps = {
			title: 'myTitle',
			onEdit: jest.fn(),
		};
	});
	it('should render', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} subTitle="mySubTitle" />);
		expect(wrapper.find(Action)).toHaveLength(0);
		expect(wrapper.find('h1')).toHaveLength(1);
		expect(wrapper.find('button')).toHaveLength(0);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render with title', () => {
		const wrapper = shallow(<TitleSubHeader {...defaultProps} />);
		expect(wrapper.find('h1').getElement().props.children).toEqual('myTitle');
	});
});

describe('SubTitle', () => {
	let defaultProps;
	beforeEach(() => {
		defaultProps = {
			subTitle: 'mySubTitle',
		};
	});

	it('should render', () => {
		const wrapper = shallow(<SubTitle {...defaultProps} />);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render in loading mode', () => {
		const wrapper = shallow(<SubTitle subTitleLoading />);
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render with a custom subtitle', () => {
		const wrapper = shallow(
			<SubTitle
				{...defaultProps}
				subTitleAs={({ subTitle }) => <Label className="label-info">{subTitle}</Label>}
			/>,
		);
		expect(wrapper.find(Label)).not.toBe(null);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
