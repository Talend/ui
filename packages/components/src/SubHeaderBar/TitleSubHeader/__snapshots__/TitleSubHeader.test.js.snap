// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleSubHeader should render 1`] = `
<div
  className="theme-tc-subheader-details tc-subheader-details"
>
  <Icon
    className="theme-tc-subheader-details-icon tc-subheader-details-icon"
    name="myIconId"
  />
  <div
    className="theme-tc-subheader-details-text tc-subheader-details-text"
  >
    <div
      className="theme-tc-subheader-details-text-title tc-subheader-details-text-title"
    >
      <h1
        className="theme-tc-subheader-details-text-title-wording tc-subheader-details-text-title-wording"
      >
        myTitle
      </h1>
    </div>
  </div>
</div>
`;

exports[`TitleSubHeader should render 2`] = `
<div
  className="theme-tc-subheader-details tc-subheader-details"
>
  <div
    className="theme-tc-subheader-details-text tc-subheader-details-text"
  >
    <div
      className="theme-tc-subheader-details-text-title tc-subheader-details-text-title"
    >
      <h1
        className="theme-tc-subheader-details-text-title-wording tc-subheader-details-text-title-wording"
      >
        myTitle
      </h1>
    </div>
    <small
      className="theme-tc-subheader-details-text-subtitle tc-subheader-details-text-subtitle"
    >
      mySubTitle
    </small>
  </div>
</div>
`;
